contracts/pcv/PCVSwapperUniswap.sol

```diff
@@ -0,0 +1,247 @@
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import "./IPCVSwapper.sol";
+import "../refs/OracleRef.sol";
+import "../utils/Timed.sol";
+import "../external/UniswapV2Library.sol";
+import "@openzeppelin/contracts/math/Math.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "@uniswap/v2-periphery/contracts/interfaces/IWETH.sol";
+import "@uniswap/v2-core/contracts/interfaces/IUniswapV2Pair.sol";
+
+/// @title implementation for PCV Swapper that swaps ERC20 tokens on Uniswap
+/// @author eswak
+contract PCVSwapperUniswap is IPCVSwapper, OracleRef, Timed {
+    using SafeERC20 for ERC20;
+    using Decimal for Decimal.D256;
+    using SafeMath for uint256;
+
+    /// @notice the token to spend on swap (outbound)
+    address public immutable override tokenSpent;
+    /// @notice the token to receive on swap (inbound)
+    address public immutable override tokenReceived;
+    /// @notice the address that will receive the inbound tokens
+    address public override tokenReceivingAddress;
+    /// @notice the maximum amount of tokens to spend on every swap
+    uint256 public maxSpentPerSwap;
+    /// @notice should we use (1 / oraclePrice) instead of oraclePrice ?
+    bool public invertOraclePrice;
+    /// @notice the incentive for calling swap() function, in FEI
+    uint256 public immutable swapIncentiveAmount;
+    /// @notice the maximum amount of slippage vs oracle price
+    uint256 public maximumSlippageBasisPoints;
+    uint256 public constant BASIS_POINTS_GRANULARITY = 10_000;
+
+    /// @notice Uniswap pair to swap on
+    IUniswapV2Pair public immutable pair;
+
+    // solhint-disable-next-line var-name-mixedcase
+    IWETH public immutable WETH;
+
+    constructor(
+        address _core,
+        IUniswapV2Pair _pair,
+        // solhint-disable-next-line var-name-mixedcase
+        IWETH _WETH,
+        address _oracle,
+        uint256 _swapFrequency,
+        address _tokenSpent,
+        address _tokenReceived,
+        address _tokenReceivingAddress,
+        uint256 _maxSpentPerSwap,
+        uint256 _maximumSlippageBasisPoints,
+        bool _invertOraclePrice,
+        uint256 _swapIncentiveAmount
+    ) public OracleRef(_core, _oracle) Timed(_swapFrequency) {
+        pair = _pair;
+        WETH = _WETH;
+        tokenSpent = _tokenSpent;
+        tokenReceived = _tokenReceived;
+        tokenReceivingAddress = _tokenReceivingAddress;
+        maxSpentPerSwap = _maxSpentPerSwap;
+        maximumSlippageBasisPoints = _maximumSlippageBasisPoints;
+        invertOraclePrice = _invertOraclePrice;
+        swapIncentiveAmount = _swapIncentiveAmount;
+
+        // start timer
+        _initTimed();
+    }
+
+    /// @notice All received ETH is wrapped to WETH
+    receive() external payable {
+      if (msg.sender != address(WETH)) {
+        WETH.deposit{value: msg.value}();
+      }
```

Comment:
To solve 4.2 and 4.5, we should break this out into a public method that wraps all held ETH in the contract and remove it from the receive function